<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Conformal Prediction for cell type annotation • scConform</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Conformal Prediction for cell type annotation">
<meta property="og:description" content="scConform">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scConform</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/scConform.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ccb-hms/scConform/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Conformal Prediction for cell type
annotation</h1>
                        <h4 data-toc-skip class="author">Daniela
Corbetta</h4>
            <address class="author_afil">
      Department of Statistical Sciences, University of
Padova<br><a class="author_email" href="mailto:#"></a><a href="mailto:daniela.corbetta@phd.unipd.it" class="email">daniela.corbetta@phd.unipd.it</a>
      </address>
                  
      
      <small class="dont-index">Source: <a href="https://github.com/ccb-hms/scConform/blob/HEAD/vignettes/scConform.Rmd" class="external-link"><code>vignettes/scConform.Rmd</code></a></small>
      <div class="hidden name"><code>scConform.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Cell type annotation is a crucial step in bioinformatic research. A
multitude of annotated datasets are now readily accessible, providing
valuable references for annotating cells in unannotated datasets
originating from similar tissues. Typically, a model is chosen and
trained on the reference data to predict the label of a new, unannotated
cell in the query dataset.</p>
<p>These methods commonly provide point predictions of the cell label,
along with estimated probabilities or scores assigned to each cell type
in the reference dataset. However, relying solely on point predictions
can be problematic when the corresponding estimated probability is low,
leading to unreliable classification. To address this issue, we propose
to exploit conformal inference to return prediction sets that include
multiple labels, with the set size reflecting the confidence in the
point prediction.</p>
<p>The package <code>scConform</code> implements two methods to achieve
this:</p>
<ol style="list-style-type: decimal">
<li><p>The first method uses split conformal inference (as a reference,
see for example section 1 of <a href="https://arxiv.org/abs/2107.07511" class="external-link">Angelopoulus and Bates,
2022</a>). The output is a prediction set, <span class="math inline">\(C(X_{new})\)</span>, that contains some of the
cell types present in the reference data. Let <span class="math inline">\(Y_{new}\)</span> be the true cell type of a new
cell in the query dataset. Then, <span class="math inline">\(C(X_{new})\)</span> satisfies <span class="math inline">\(P(Y_{new} \in C(X_{new})) \geq 1-\alpha\)</span>,
where <span class="math inline">\(\alpha\)</span> is a user-chosen error
level.</p></li>
<li><p>The second method is based on conformal risk control (<a href="https://arxiv.org/abs/2208.02814" class="external-link">Angelopoulus et al., 2023</a>).
It considers the inherent relationships among cell types that are
encoded as graph-structured constraints available through the cell
ontology. The final output is a prediction set aligned with the cell
ontology structure. Specifically, the prediction set will include all
the children of an ancestor of the predicted class. The more unsure the
point prediction, the broader the classification. Also this method
satisfies <span class="math inline">\(P(Y_{new} \in C(X_{new})) \geq
1-\alpha\)</span>.</p></li>
</ol>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>To install it</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/version.html" class="external-link">version</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="st">"3.20"</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"scConform"</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="fu">spdl</span><span class="fu">::</span><span class="fu">info</span><span class="op">(</span></span>
<span>        <span class="st">"'scConform' requires Bioconductor version 3.20 or later, "</span>,</span>
<span>        <span class="st">"installing development version from Github"</span></span>
<span>    <span class="op">)</span></span>
<span>    <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"ccb-hms/scConform"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="preliminaries">Preliminaries<a class="anchor" aria-label="anchor" href="#preliminaries"></a>
</h2>
<div class="section level3">
<h3 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ccb-hms/scConform" class="external-link">scConform</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.stat.auckland.ac.nz/~yee/VGAM/" class="external-link">VGAM</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/vjcitn/ontoProc" class="external-link">ontoProc</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ccb-hms/MerfishData" class="external-link">MerfishData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r.igraph.org/" class="external-link">igraph</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">`%notin%`</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/funprog.html" class="external-link">Negate</a></span><span class="op">(</span><span class="va">`%in%`</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="load-data-and-cell-ontology">Load data and cell ontology<a class="anchor" aria-label="anchor" href="#load-data-and-cell-ontology"></a>
</h3>
<p>As an example, we’ll use the mouse ileum Merfish data from the
<code>MerfishData</code> Bioconductor package, segmented with Baysor.
For the purpose of this vignette, we’ll treat the data as if they were
single-cell data, discarding the spatial information. This is a toy
example in which reference and query data will be a subset of the same
dataset. In real life examples, data integration is a necessary first
step.</p>
<p>Next, load the cell ontology trough the <code>ontoProc</code>
Bioconductor package.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load data</span></span>
<span><span class="va">spe_baysor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MerfishData/man/MouseIleumPetukhov2021.html" class="external-link">MouseIleumPetukhov2021</a></span><span class="op">(</span></span>
<span>    segmentation <span class="op">=</span> <span class="st">"baysor"</span>,</span>
<span>    use.images <span class="op">=</span> <span class="cn">FALSE</span>, use.polygons <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Load ontology</span></span>
<span><span class="va">cl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ontoProc/man/getOnto.html" class="external-link">getOnto</a></span><span class="op">(</span><span class="st">"cellOnto"</span>, <span class="st">"2023"</span><span class="op">)</span></span>
<span><span class="co">#&gt; loading from cache</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="build-the-ontology">Build the ontology<a class="anchor" aria-label="anchor" href="#build-the-ontology"></a>
</h3>
<p>The object <code>cl</code> contains information regarding all the
relationships among all the known cell types. The first step is to
filter the ontology and only keep the parts related to the cell types
that are present in the data.</p>
<p>This dataset does not provide annotations with the cell ontology
tags, so we need to browse the ontology to find the interesting
tags.</p>
<p>For more information on how to restrict the cell ontology, refer to
the <a href="https://www.bioconductor.org/packages/release/bioc/html/ontoProc.html" class="external-link">ontoProc</a>
documentation.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tags</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"CL:0009022"</span>, <span class="co"># Stromal</span></span>
<span>    <span class="st">"CL:0000236"</span>, <span class="co"># B cell</span></span>
<span>    <span class="st">"CL:0009080"</span>, <span class="co"># Tuft</span></span>
<span>    <span class="st">"CL:1000411"</span>, <span class="co"># Endothelial</span></span>
<span>    <span class="st">"CL:1000335"</span>, <span class="co"># Enterocyte</span></span>
<span>    <span class="st">"CL:1000326"</span>, <span class="co"># Goblet</span></span>
<span>    <span class="st">"CL:0002088"</span>, <span class="co"># ICC</span></span>
<span>    <span class="st">"CL:0009007"</span>, <span class="co"># Macrophage + DC</span></span>
<span>    <span class="st">"CL:1000343"</span>, <span class="co"># Paneth</span></span>
<span>    <span class="st">"CL:0000669"</span>, <span class="co"># Pericyte</span></span>
<span>    <span class="st">"CL:1000278"</span>, <span class="co"># Smooth Muscle</span></span>
<span>    <span class="st">"CL:0009017"</span>, <span class="co"># Stem + TA</span></span>
<span>    <span class="st">"CL:0000492"</span>, <span class="co"># T (CD4+)</span></span>
<span>    <span class="st">"CL:0000625"</span>, <span class="co"># T (CD8+)</span></span>
<span>    <span class="st">"CL:0017004"</span> <span class="co"># Telocyte</span></span>
<span><span class="op">)</span></span>
<span><span class="va">opi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://r.igraph.org/reference/graph_from_graphnel.html" class="external-link">graph_from_graphnel</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ontoProc/man/onto_plot2.html" class="external-link">onto_plot2</a></span><span class="op">(</span><span class="va">cl</span>, <span class="va">tags</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>In the <code>opi</code> object, there are also instances from other
ontologies (CARO and BFO) that need to be removed.</p>
<p>Moreover, the names of the leaf nodes of the ontology must correspond
to the names used for the annotation.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Delete CARO and BFO instances</span></span>
<span><span class="va">sel_ver</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"CARO"</span>, <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"BFO"</span>, <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">opi1</span> <span class="op">&lt;-</span> <span class="va">opi</span> <span class="op">-</span> <span class="va">sel_ver</span></span>
<span></span>
<span><span class="co">## Rename vertex to match annotations</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi1</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"CL:0000236"</span>, <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi1</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"B cell"</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi1</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"CL:1000411"</span>, <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi1</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Endothelial"</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi1</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"CL:1000335"</span>, <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi1</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Enterocyte"</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi1</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"CL:1000326"</span>, <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi1</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Goblet"</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi1</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"CL:0002088"</span>, <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi1</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"ICC"</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi1</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"CL:0009007"</span>, <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi1</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Macrophage + DC"</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi1</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"CL:1000343"</span>, <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi1</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Paneth"</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi1</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"CL:0000669"</span>, <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi1</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Pericyte"</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi1</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"CL:1000278"</span>, <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi1</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Smooth Muscle"</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi1</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"CL:0009017"</span>, <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi1</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Stem + TA"</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi1</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"CL:0009022"</span>, <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi1</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Stromal"</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi1</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"CL:0000492"</span>, <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi1</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"T (CD4+)"</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi1</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"CL:0000625"</span>, <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi1</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"T (CD8+)"</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi1</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"CL:0017004"</span>, <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi1</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Telocyte"</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi1</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"CL:0009080"</span>, <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi1</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Tuft"</span></span>
<span></span>
<span><span class="co">## Add the edge from connective tissue cell and telocyte and delete redundant</span></span>
<span><span class="co">## nodes</span></span>
<span><span class="va">opi1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://r.igraph.org/reference/add_edges.html" class="external-link">add_edges</a></span><span class="op">(</span><span class="va">opi1</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"connective\ntissue cell\nCL:0002320"</span>, <span class="st">"Telocyte"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://r.igraph.org/reference/as_graphnel.html" class="external-link">as_graphnel</a></span><span class="op">(</span><span class="va">opi1</span><span class="op">)</span></span>
<span><span class="va">opi2</span> <span class="op">&lt;-</span> <span class="va">opi1</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"somatic\ncell\nCL:0002371"</span>, <span class="st">"contractile\ncell\nCL:0000183"</span>,</span>
<span>    <span class="st">"native\ncell\nCL:0000003"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi2</span><span class="op">)</span><span class="op">$</span><span class="va">name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/trimws.html" class="external-link">trimws</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"CL:.*|\\n"</span>, <span class="st">" "</span>, <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi2</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gr1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://r.igraph.org/reference/as_graphnel.html" class="external-link">as_graphnel</a></span><span class="op">(</span><span class="va">opi2</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Plot the final ontology</span></span>
<span><span class="va">attrs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>node <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="st">"box"</span>, fontsize <span class="op">=</span> <span class="fl">15</span>, fixedsize <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">gr1</span>, attrs <span class="op">=</span> <span class="va">attrs</span><span class="op">)</span></span></code></pre></div>
<p><img src="scConform_files/figure-html/build-ontology-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="preprocess-data">Preprocess data<a class="anchor" aria-label="anchor" href="#preprocess-data"></a>
</h3>
<p>Modify the colData variable <code>leiden_final</code> to unify B
cells and enterocytes</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe_baysor</span><span class="op">$</span><span class="va">cell_type</span> <span class="op">&lt;-</span> <span class="va">spe_baysor</span><span class="op">$</span><span class="va">leiden_final</span></span>
<span><span class="va">spe_baysor</span><span class="op">$</span><span class="va">cell_type</span><span class="op">[</span><span class="va">spe_baysor</span><span class="op">$</span><span class="va">cell_type</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"B (Follicular, Circulating)"</span>,</span>
<span>    <span class="st">"B (Plasma)"</span></span>
<span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"B cell"</span></span>
<span><span class="va">spe_baysor</span><span class="op">$</span><span class="va">cell_type</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"Enterocyte"</span>, <span class="va">spe_baysor</span><span class="op">$</span><span class="va">cell_type</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Enterocyte"</span></span>
<span><span class="va">spe_baysor</span> <span class="op">&lt;-</span> <span class="va">spe_baysor</span><span class="op">[</span>, <span class="va">spe_baysor</span><span class="op">$</span><span class="va">cell_type</span> <span class="op">%notin%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"Removed"</span>,</span>
<span>    <span class="st">"Myenteric Plexus"</span></span>
<span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">spe_baysor</span></span>
<span><span class="co">#&gt; class: SpatialExperiment </span></span>
<span><span class="co">#&gt; dim: 241 5163 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(2): counts molecules</span></span>
<span><span class="co">#&gt; rownames(241): Acsl1 Acta2 ... Vcan Vim</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames: NULL</span></span>
<span><span class="co">#&gt; colData names(8): n_transcripts density ... sample_id cell_type</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : x y</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># See frequencies of cell types</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">spe_baysor</span><span class="op">$</span><span class="va">cell_type</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;          B cell     Endothelial      Enterocyte          Goblet             ICC </span></span>
<span><span class="co">#&gt;             536             231            1257             299              31 </span></span>
<span><span class="co">#&gt; Macrophage + DC          Paneth        Pericyte   Smooth Muscle       Stem + TA </span></span>
<span><span class="co">#&gt;             427             328             102             428             580 </span></span>
<span><span class="co">#&gt;         Stromal        T (CD4+)        T (CD8+)        Telocyte            Tuft </span></span>
<span><span class="co">#&gt;             489             197             125             115              18</span></span></code></pre></div>
<p>Now randomly split the data into reference and query data.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1636</span><span class="op">)</span></span>
<span><span class="va">ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">spe_baysor</span><span class="op">)</span><span class="op">)</span>, <span class="fl">600</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">spe_ref</span> <span class="op">&lt;-</span> <span class="va">spe_baysor</span><span class="op">[</span>, <span class="va">ref</span><span class="op">]</span></span>
<span><span class="va">spe_query</span> <span class="op">&lt;-</span> <span class="va">spe_baysor</span><span class="op">[</span>, <span class="op">-</span><span class="va">ref</span><span class="op">]</span></span>
<span><span class="co"># Reference data</span></span>
<span><span class="va">spe_ref</span></span>
<span><span class="co">#&gt; class: SpatialExperiment </span></span>
<span><span class="co">#&gt; dim: 241 600 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(2): counts molecules</span></span>
<span><span class="co">#&gt; rownames(241): Acsl1 Acta2 ... Vcan Vim</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames: NULL</span></span>
<span><span class="co">#&gt; colData names(8): n_transcripts density ... sample_id cell_type</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : x y</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Query data</span></span>
<span><span class="va">spe_query</span></span>
<span><span class="co">#&gt; class: SpatialExperiment </span></span>
<span><span class="co">#&gt; dim: 241 4563 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(2): counts molecules</span></span>
<span><span class="co">#&gt; rownames(241): Acsl1 Acta2 ... Vcan Vim</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames: NULL</span></span>
<span><span class="co">#&gt; colData names(8): n_transcripts density ... sample_id cell_type</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : x y</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="fit-a-classification-model">Fit a classification model<a class="anchor" aria-label="anchor" href="#fit-a-classification-model"></a>
</h3>
<p>We now need to build a classification model. To this aim, we’ll draw
a random sample of 300 cells from the reference data, and train a
multinomial model on those. As explanatory variables we’ll use the 50
most variable genes. For this step, every statistical model or machine
learning method can be used, as long as it provides also estimated
probabilities for each class.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Randomly select 300 cells</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1704</span><span class="op">)</span></span>
<span><span class="va">train</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">spe_ref</span><span class="op">)</span><span class="op">)</span>, <span class="fl">300</span><span class="op">)</span></span>
<span><span class="co"># Training data</span></span>
<span><span class="va">spe_train</span> <span class="op">&lt;-</span> <span class="va">spe_ref</span><span class="op">[</span>, <span class="va">train</span><span class="op">]</span></span>
<span><span class="va">spe_train</span></span>
<span><span class="co">#&gt; class: SpatialExperiment </span></span>
<span><span class="co">#&gt; dim: 241 300 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(2): counts molecules</span></span>
<span><span class="co">#&gt; rownames(241): Acsl1 Acta2 ... Vcan Vim</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames: NULL</span></span>
<span><span class="co">#&gt; colData names(8): n_transcripts density ... sample_id cell_type</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : x y</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># get HVGs</span></span>
<span><span class="va">spe_train</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">spe_train</span><span class="op">)</span></span>
<span><span class="va">v</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">spe_train</span><span class="op">)</span></span>
<span><span class="va">hvg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">v</span>, n <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract counts and convert data into a data.frame format</span></span>
<span><span class="va">df_train</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">spe_train</span><span class="op">[</span><span class="va">hvg</span>, <span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">df_train</span><span class="op">$</span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="va">spe_train</span><span class="op">$</span><span class="va">cell_type</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">df_train</span><span class="op">$</span><span class="va">Y</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;          B cell     Endothelial      Enterocyte          Goblet             ICC </span></span>
<span><span class="co">#&gt;              25              10              68              19               2 </span></span>
<span><span class="co">#&gt; Macrophage + DC          Paneth        Pericyte   Smooth Muscle       Stem + TA </span></span>
<span><span class="co">#&gt;              28              17               4              22              48 </span></span>
<span><span class="co">#&gt;         Stromal        T (CD4+)        T (CD8+)        Telocyte            Tuft </span></span>
<span><span class="co">#&gt;              34              13               5               4               1</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Fit multinomial model</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/VGAM/man/vglm.html" class="external-link">vglm</a></span><span class="op">(</span><span class="va">Y</span> <span class="op">~</span> <span class="va">.</span>,</span>
<span>    family <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/VGAM/man/multinomial.html" class="external-link">multinomial</a></span><span class="op">(</span>refLevel <span class="op">=</span> <span class="st">"B cell"</span><span class="op">)</span>,</span>
<span>    data <span class="op">=</span> <span class="va">df_train</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="method-1-standard-conformal-inference">Method 1: standard conformal inference<a class="anchor" aria-label="anchor" href="#method-1-standard-conformal-inference"></a>
</h2>
<div class="section level3">
<h3 id="algorithm">Algorithm<a class="anchor" aria-label="anchor" href="#algorithm"></a>
</h3>
<p>As anticipated in the introduction, the first method uses split
conformal inference to build the prediction sets. It requires to split
the reference data into two subsets:</p>
<ul>
<li>training set: portion of the reference data to be used to build a
classifier <span class="math inline">\(\hat{f}\)</span> (the multinomial
model in our example)</li>
<li>calibration set: portion of the reference data to be used to
calibrate the procedure.</li>
</ul>
<p>Given <span class="math inline">\(\hat{f}\)</span> and the
calibration data, the objective of conformal inference is to build, for
a new non-annotated cell <span class="math inline">\(X_{new}\)</span>, a
prediction set <span class="math inline">\(C(X_{new})\)</span> that
satisfies</p>
<p><span class="math display">\[\begin{equation}
P(Y_{new} \in C(X_{new})) \geq 1-\alpha,
(\#eq:cov)
\end{equation}\]</span></p>
<p>where <span class="math inline">\(Y_{new}\)</span> is the true label
of the new cell and <span class="math inline">\(\alpha\)</span> is a
user-chosen error level. The algorithm for split conformal inference is
the following:</p>
<ol style="list-style-type: decimal">
<li><p>For the data in the calibration set, <span class="math inline">\((X_1, Y_1), \dots, (X_n, Y_n),\)</span> obtain the
conformal scores <span class="math inline">\(s_1=1-\hat{f}(X_i)_{Y_i}\)</span> (i.e. 1 minus
the estimated probability that the model is assigning to the true label
of the <span class="math inline">\(i\)</span>-th cell). These scores
will be high when the model is assigning a small probability to the true
class, and low otherwise.</p></li>
<li><p>Obtain <span class="math inline">\(\hat{q}\)</span> as the <span class="math inline">\(\lceil(1-\alpha)(n+1)\rceil/n\)</span> empirical
quantile of the conformal scores.</p></li>
<li><p>Finally, for a new cell <span class="math inline">\(X_{new}\)</span>, build a prediction set by
including all the classes or which the estimated probability is higher
than <span class="math inline">\(1-\hat{q}\)</span>: <span class="math display">\[
C(X_{new}) = \left\{y: \hat{f}(X_{n+1})_y\geq 1-\hat{q}\right\}
\]</span></p></li>
</ol>
</div>
<div class="section level3">
<h3 id="obtain-prediction-matrices">Obtain prediction matrices<a class="anchor" aria-label="anchor" href="#obtain-prediction-matrices"></a>
</h3>
<p>The first step to build conformal prediction sets is to obtain
matrices with the estimated probabilities for each cell type for cells
in the calibration data and in the query data. Each row of the matrix
corresponds to a particular cell, while each row to a different cell
type. The entry <span class="math inline">\(p_{i,j}\)</span> of the
matrix indicates the estimated probability that the cell <span class="math inline">\(i\)</span> is of type <span class="math inline">\(j\)</span>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe_cal</span> <span class="op">&lt;-</span> <span class="va">spe_ref</span><span class="op">[</span>, <span class="op">-</span><span class="va">train</span><span class="op">]</span></span>
<span><span class="co"># Prediction matrix for calibration data</span></span>
<span><span class="va">df_cal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">spe_cal</span><span class="op">[</span><span class="va">hvg</span>, <span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p_cal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">fit</span>, newdata <span class="op">=</span> <span class="va">df_cal</span>, type <span class="op">=</span> <span class="st">"response"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">p_cal</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;   B cell Endothelial Enterocyte Goblet   ICC Macrophage + DC Paneth Pericyte</span></span>
<span><span class="co">#&gt; 1  0.001       0.000      0.023  0.000 0.000           0.561  0.000    0.000</span></span>
<span><span class="co">#&gt; 2  0.000       0.000      0.000  0.000 0.000           0.000  0.000    0.000</span></span>
<span><span class="co">#&gt; 3  0.018       0.000      0.041  0.000 0.001           0.923  0.000    0.001</span></span>
<span><span class="co">#&gt; 4  0.001       0.001      0.283  0.035 0.001           0.003  0.017    0.000</span></span>
<span><span class="co">#&gt; 5  0.000       0.000      1.000  0.000 0.000           0.000  0.000    0.000</span></span>
<span><span class="co">#&gt; 6  0.000       0.000      0.002  0.000 0.000           0.000  0.141    0.000</span></span>
<span><span class="co">#&gt;   Smooth Muscle Stem + TA Stromal T (CD4+) T (CD8+) Telocyte  Tuft</span></span>
<span><span class="co">#&gt; 1         0.002     0.000   0.407    0.000    0.000    0.006 0.000</span></span>
<span><span class="co">#&gt; 2         0.000     0.000   0.000    0.001    0.999    0.000 0.000</span></span>
<span><span class="co">#&gt; 3         0.005     0.007   0.001    0.001    0.000    0.000 0.001</span></span>
<span><span class="co">#&gt; 4         0.004     0.161   0.488    0.001    0.002    0.003 0.001</span></span>
<span><span class="co">#&gt; 5         0.000     0.000   0.000    0.000    0.000    0.000 0.000</span></span>
<span><span class="co">#&gt; 6         0.000     0.856   0.000    0.000    0.000    0.000 0.000</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Prediction matrix for query data</span></span>
<span><span class="va">df_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">spe_query</span><span class="op">[</span><span class="va">hvg</span>, <span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">fit</span>, newdata <span class="op">=</span> <span class="va">df_test</span>, type <span class="op">=</span> <span class="st">"response"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">p_test</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;   B cell Endothelial Enterocyte Goblet ICC Macrophage + DC Paneth Pericyte</span></span>
<span><span class="co">#&gt; 1      0       0.627          0      0   0               0      0        0</span></span>
<span><span class="co">#&gt; 2      0       0.000          0      0   0               0      0        0</span></span>
<span><span class="co">#&gt; 3      0       0.001          0      0   0               0      0        0</span></span>
<span><span class="co">#&gt; 4      0       0.000          0      0   0               0      0        0</span></span>
<span><span class="co">#&gt; 5      0       0.000          0      0   0               0      0        0</span></span>
<span><span class="co">#&gt; 6      0       0.000          0      0   0               0      0        0</span></span>
<span><span class="co">#&gt;   Smooth Muscle Stem + TA Stromal T (CD4+) T (CD8+) Telocyte Tuft</span></span>
<span><span class="co">#&gt; 1         0.000         0   0.001    0.000        0    0.372    0</span></span>
<span><span class="co">#&gt; 2         0.997         0   0.000    0.000        0    0.003    0</span></span>
<span><span class="co">#&gt; 3         0.021         0   0.002    0.001        0    0.975    0</span></span>
<span><span class="co">#&gt; 4         1.000         0   0.000    0.000        0    0.000    0</span></span>
<span><span class="co">#&gt; 5         0.930         0   0.069    0.000        0    0.001    0</span></span>
<span><span class="co">#&gt; 6         1.000         0   0.000    0.000        0    0.000    0</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="obtain-conformal-prediction-sets">Obtain conformal prediction sets<a class="anchor" aria-label="anchor" href="#obtain-conformal-prediction-sets"></a>
</h3>
<p>We can now directly call the <code>getPredictionSet</code> function
by using as input the prediction matrices for the calibration and the
query dataset. In this case, the output of the function will be a list
whose elements are the prediction sets for each query cell.</p>
<p>By setting <code>follow_ontology=FALSE</code>, we are asking the
function to return prediction sets obtained via split conformal
inference. The parameter <code>onto</code> is not necessary with this
method, but if the ontology is provided then it will be used by the
function to retrieve the considered cell types. Alternatively, we can
explicitly provide the labels with the parameter <code>labels</code>.
The parameter <code>alpha</code> indicates the allowed miscoverage
percentage (see @ref(eq:cov)). For example, if we set
<code>alpha=0.1</code>, it means that at most 10% of the prediction sets
we obtain will not include the true label.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">p_test</span><span class="op">)</span></span>
<span><span class="va">sets</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getPredictionSets.html">getPredictionSets</a></span><span class="op">(</span></span>
<span>    x_query <span class="op">=</span> <span class="va">p_test</span>,</span>
<span>    x_cal <span class="op">=</span> <span class="va">p_cal</span>,</span>
<span>    y_cal <span class="op">=</span> <span class="va">spe_cal</span><span class="op">$</span><span class="va">cell_type</span>,</span>
<span>    alpha <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>    follow_ontology <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    labels <span class="op">=</span> <span class="va">labels</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># See the first six prediction sets</span></span>
<span><span class="va">sets</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; [1] "Endothelial" "Telocyte"   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt; [1] "Smooth Muscle"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[3]]</span></span>
<span><span class="co">#&gt; [1] "Smooth Muscle" "Telocyte"     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[4]]</span></span>
<span><span class="co">#&gt; [1] "Smooth Muscle"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[5]]</span></span>
<span><span class="co">#&gt; [1] "Smooth Muscle" "Stromal"      </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[6]]</span></span>
<span><span class="co">#&gt; [1] "Smooth Muscle"</span></span></code></pre></div>
<p>Since in this example the cell labels are available, we can
explicitly check that @ref(eq:cov) is satisfied by computing the
proportion of sets that include the true label. We set
<code>alpha=0.1</code>, so we expect this proportion to be higher than
0.9.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Check coverage</span></span>
<span><span class="va">cvg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">sets</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">sets</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">cvg</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">spe_query</span><span class="op">$</span><span class="va">cell_type</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">sets</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">cvg</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.9035722</span></span></code></pre></div>
<p>As an alternative, we can provide as input a
<code>SingleCellExperiment</code> object. In this case, it needs to have
the estimated probabilities for each cell type in the
<code>colData</code>. The names of these <code>colData</code> have to
correspond to the names of the leaf nodes in the ontology.</p>
<p>The output will be a <code>SingleCellExperiment</code> object with a
new variable in the <code>colData</code> containing the prediction sets.
The name of this variable can be assigned trough the
<code>pr_name</code> parameter.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Retrieve labels as leaf nodes of the ontology</span></span>
<span><span class="va">labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">opi2</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">[</span><span class="fu"><a href="https://r.igraph.org/reference/degree.html" class="external-link">degree</a></span><span class="op">(</span><span class="va">opi2</span>, mode <span class="op">=</span> <span class="st">"out"</span><span class="op">)</span> <span class="op">==</span> <span class="fl">0</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Create corresponding colData</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="va">labels</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">spe_cal</span><span class="op">)</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">p_cal</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">spe_query</span><span class="op">)</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">p_test</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Create prediction sets</span></span>
<span><span class="va">spe_query</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getPredictionSets.html">getPredictionSets</a></span><span class="op">(</span></span>
<span>    x_query <span class="op">=</span> <span class="va">spe_query</span>,</span>
<span>    x_cal <span class="op">=</span> <span class="va">spe_cal</span>,</span>
<span>    y_cal <span class="op">=</span> <span class="va">spe_cal</span><span class="op">$</span><span class="va">cell_type</span>,</span>
<span>    alpha <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>    follow_ontology <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    pr_name <span class="op">=</span> <span class="st">"pred_set"</span>,</span>
<span>    labels <span class="op">=</span> <span class="va">labels</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># See the new variable pred_set into the colData</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">spe_query</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 6 rows and 24 columns</span></span>
<span><span class="co">#&gt;   n_transcripts   density elongation      area avg_confidence  leiden_final</span></span>
<span><span class="co">#&gt;       &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt;      &lt;numeric&gt;   &lt;character&gt;</span></span>
<span><span class="co">#&gt; 1            39   0.02159      5.082      1806         0.8647   Endothelial</span></span>
<span><span class="co">#&gt; 2           165   0.02016      1.565      8186         0.9528 Smooth Muscle</span></span>
<span><span class="co">#&gt; 3           139   0.02279      1.820      6100         0.9762 Smooth Muscle</span></span>
<span><span class="co">#&gt; 4            80   0.01828      1.546      4376         0.9076 Smooth Muscle</span></span>
<span><span class="co">#&gt; 5            75   0.02479      3.475      3025         0.8952 Smooth Muscle</span></span>
<span><span class="co">#&gt; 6           167   0.02409      1.265      6932         0.9618 Smooth Muscle</span></span>
<span><span class="co">#&gt;     sample_id     cell_type     Stromal      B cell        Tuft Endothelial</span></span>
<span><span class="co">#&gt;   &lt;character&gt;   &lt;character&gt;   &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; 1       ileum   Endothelial 1.06770e-03 4.62746e-06 1.09285e-06 6.26716e-01</span></span>
<span><span class="co">#&gt; 2       ileum Smooth Muscle 2.48790e-07 2.11623e-12 1.71646e-09 4.27276e-13</span></span>
<span><span class="co">#&gt; 3       ileum Smooth Muscle 1.82261e-03 1.05365e-06 3.42260e-05 1.45712e-03</span></span>
<span><span class="co">#&gt; 4       ileum Smooth Muscle 4.46595e-07 8.68709e-12 1.84782e-10 1.48018e-11</span></span>
<span><span class="co">#&gt; 5       ileum Smooth Muscle 6.87351e-02 5.40229e-05 3.80670e-05 1.56797e-04</span></span>
<span><span class="co">#&gt; 6       ileum Smooth Muscle 6.47366e-15 2.80039e-21 2.51182e-16 3.76540e-21</span></span>
<span><span class="co">#&gt;    Enterocyte      Goblet         ICC Macrophage + DC      Paneth    Pericyte</span></span>
<span><span class="co">#&gt;     &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;       &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; 1 6.23838e-09 2.16089e-06 1.04065e-04     8.69702e-11 2.46469e-08 1.25208e-06</span></span>
<span><span class="co">#&gt; 2 3.65028e-15 5.45567e-19 3.26311e-10     3.71068e-15 1.04596e-11 7.83728e-11</span></span>
<span><span class="co">#&gt; 3 7.62737e-11 1.02715e-10 1.30690e-04     4.16898e-09 1.48191e-07 2.69325e-06</span></span>
<span><span class="co">#&gt; 4 3.06169e-10 4.21061e-16 8.73496e-10     1.02177e-11 3.33635e-09 1.75309e-11</span></span>
<span><span class="co">#&gt; 5 9.53445e-06 1.14109e-06 1.68140e-05     3.24168e-06 3.04576e-04 4.81784e-06</span></span>
<span><span class="co">#&gt; 6 1.27740e-21 9.27025e-38 7.72260e-16     3.18369e-21 1.04362e-26 5.48492e-22</span></span>
<span><span class="co">#&gt;   Smooth Muscle   Stem + TA    T (CD4+)    T (CD8+)    Telocyte</span></span>
<span><span class="co">#&gt;       &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; 1   3.17594e-12 1.94285e-04 2.61437e-07 1.10020e-05 3.71897e-01</span></span>
<span><span class="co">#&gt; 2   9.97273e-01 4.23329e-18 3.08082e-10 4.01534e-10 2.72648e-03</span></span>
<span><span class="co">#&gt; 3   2.11629e-02 5.03876e-11 5.90176e-04 2.05918e-05 9.74778e-01</span></span>
<span><span class="co">#&gt; 4   9.99986e-01 5.59105e-10 4.55175e-09 9.09174e-10 1.35576e-05</span></span>
<span><span class="co">#&gt; 5   9.29946e-01 5.76652e-07 5.48474e-05 1.91186e-05 6.55575e-04</span></span>
<span><span class="co">#&gt; 6   9.99984e-01 1.63394e-24 1.70917e-14 4.89589e-15 1.60623e-05</span></span>
<span><span class="co">#&gt;                 pred_set</span></span>
<span><span class="co">#&gt;                   &lt;list&gt;</span></span>
<span><span class="co">#&gt; 1   Endothelial,Telocyte</span></span>
<span><span class="co">#&gt; 2          Smooth Muscle</span></span>
<span><span class="co">#&gt; 3 Smooth Muscle,Telocyte</span></span>
<span><span class="co">#&gt; 4          Smooth Muscle</span></span>
<span><span class="co">#&gt; 5  Stromal,Smooth Muscle</span></span>
<span><span class="co">#&gt; 6          Smooth Muscle</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="method-2-exploit-the-cell-ontology">Method 2: exploit the cell ontology<a class="anchor" aria-label="anchor" href="#method-2-exploit-the-cell-ontology"></a>
</h2>
<div class="section level3">
<h3 id="algorithm-1">Algorithm<a class="anchor" aria-label="anchor" href="#algorithm-1"></a>
</h3>
<p>Let <span class="math inline">\(\hat{y}(x)\)</span> be the class with
the maximum estimated probability. Moreover, given a directed graph let
<span class="math inline">\(P(v)\)</span> and <span class="math inline">\(A(v)\)</span> be the set of children nodes and
ancestor nodes of <span class="math inline">\(v\)</span>, respectively.
Finally, for each node <span class="math inline">\(v\)</span> define a
score <span class="math inline">\(g(v,x)\)</span> as the sum of the
predicted probabilities of the leaf nodes that are children of <span class="math inline">\(v\)</span>. We build the sets as follows:</p>
<p><span class="math display">\[
P(v) \cup \{P(a): a\in A(\hat{y}(x)):
g(a,x)\leq\lambda \},
\]</span> where <span class="math inline">\(v:v\in A(\hat{y}(x)),
\;g(v,x)\geq\lambda\)</span> and <span class="math inline">\(v=\arg\min_{u:g(u,x)\geq\lambda}g(u,x)\)</span>.</p>
<p>In words, we start from the predicted class and we go up in the graph
until we find an ancestor of <span class="math inline">\(\hat{y}(x)\)</span> that has a score that is at
least <span class="math inline">\(\lambda\)</span> and include in the
prediction sets all its children. Then we add to this subgraph the other
ones that contain <span class="math inline">\(\hat{y}(x)\)</span> for
which the score is less than <span class="math inline">\(\lambda\)</span>. To choose <span class="math inline">\(\lambda\)</span>, we follow eq. (4) in <a href="https://arxiv.org/abs/2208.02814" class="external-link">Angelopoulus et al., 2023</a>
(2022), considering the miscoverage as loss function.</p>
</div>
<div class="section level3">
<h3 id="obtain-hierarchical-prediction-sets">Obtain hierarchical prediction sets<a class="anchor" aria-label="anchor" href="#obtain-hierarchical-prediction-sets"></a>
</h3>
<p>To switch to the hierarchical algorithm, we just need to set
<code>follow_ontology=TRUE</code> and provide an ontology to the
<code>onto</code> parameter. As in the previous case, if the input is a
<code>SingleCellExperiment</code> object then the output will be again a
<code>SingleCellExperiment</code> object with a new variable containing
the prediction sets in the colData.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe_query</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getPredictionSets.html">getPredictionSets</a></span><span class="op">(</span></span>
<span>    x_query <span class="op">=</span> <span class="va">spe_query</span>,</span>
<span>    x_cal <span class="op">=</span> <span class="va">spe_cal</span>,</span>
<span>    y_cal <span class="op">=</span> <span class="va">spe_cal</span><span class="op">$</span><span class="va">cell_type</span>,</span>
<span>    onto <span class="op">=</span> <span class="va">opi2</span>,</span>
<span>    alpha <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>    follow_ontology <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    pr_name <span class="op">=</span> <span class="st">"pred_set_hier"</span>,</span>
<span>    BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>,</span>
<span>    simplify_pred <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># See the first six prediction sets</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">spe_query</span><span class="op">$</span><span class="va">pred_set_hier</span><span class="op">)</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt;  [1] "Tuft"            "Endothelial"     "Enterocyte"      "Goblet"         </span></span>
<span><span class="co">#&gt;  [5] "ICC"             "Paneth"          "Stem + TA"       "Stromal"        </span></span>
<span><span class="co">#&gt;  [9] "B cell"          "Macrophage + DC" "Pericyte"        "Smooth Muscle"  </span></span>
<span><span class="co">#&gt; [13] "T (CD4+)"        "T (CD8+)"        "Telocyte"       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt; [1] "Smooth Muscle"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[3]]</span></span>
<span><span class="co">#&gt; [1] "Telocyte"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[4]]</span></span>
<span><span class="co">#&gt; [1] "Smooth Muscle"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[5]]</span></span>
<span><span class="co">#&gt; [1] "Smooth Muscle"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[6]]</span></span>
<span><span class="co">#&gt; [1] "Smooth Muscle"</span></span></code></pre></div>
<p>Also in this case, we can check the coverage:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Check coverage</span></span>
<span><span class="va">cvg1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">spe_query</span><span class="op">$</span><span class="va">pred_set_hier</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">spe_query</span><span class="op">$</span><span class="va">pred_set_hier</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">cvg1</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">spe_query</span><span class="op">$</span><span class="va">cell_type</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">spe_query</span><span class="op">$</span><span class="va">pred_set_hier</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">cvg1</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.8912996</span></span></code></pre></div>
<p>As an alternative, we can choose to return only the common ancestor,
instead of the entire set of predicted labels. This can be done directly
in the function by setting <code>simplify_pred=TRUE</code>.
Alternatively, we can convert the obtained prediction sets invoking the
function <code>returnCommonAncestor</code>.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe_query</span><span class="op">$</span><span class="va">pred_set_hier_simp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">vapply</a></span><span class="op">(</span></span>
<span>    <span class="va">spe_query</span><span class="op">$</span><span class="va">pred_set_hier</span>,</span>
<span>    <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="../reference/returnCommonAncestor.html">returnCommonAncestor</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">opi2</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">spe_query</span><span class="op">$</span><span class="va">pred_set_hier_simp</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "animal cell"   "Smooth Muscle" "Telocyte"      "Smooth Muscle"</span></span>
<span><span class="co">#&gt; [5] "Smooth Muscle" "Smooth Muscle"</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h2>
<p>Let’s now visualize the obtained prediction sets. The simplest way is
to just ask to color the labels included in the prediction set. Let’s
compare the prediction sets for one cell obtained with the two
methods</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotResult.html">plotResult</a></span><span class="op">(</span><span class="va">spe_query</span><span class="op">$</span><span class="va">pred_set</span><span class="op">[[</span><span class="fl">75</span><span class="op">]</span><span class="op">]</span>, <span class="va">opi2</span>,</span>
<span>    col_grad <span class="op">=</span> <span class="st">"pink"</span>, attrs <span class="op">=</span> <span class="va">attrs</span>, add_scores <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    title <span class="op">=</span> <span class="st">"Conformal Prediction set"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="scConform_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotResult.html">plotResult</a></span><span class="op">(</span><span class="va">spe_query</span><span class="op">$</span><span class="va">pred_set_hier</span><span class="op">[[</span><span class="fl">75</span><span class="op">]</span><span class="op">]</span>, <span class="va">opi2</span>,</span>
<span>    col_grad <span class="op">=</span> <span class="st">"pink"</span>, attrs <span class="op">=</span> <span class="va">attrs</span>, add_scores <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    title <span class="op">=</span> <span class="st">"Hierarchical prediction set"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="scConform_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<p>We can now instead give as input also the estimated probabilities and
a gradient. The function will color the cells according to the gradient,
with stronger colors corresponding to higher probabilities. If
<code>add_scores=TRUE</code>, then the estimated probabilities will be
added to the labels.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotResult.html">plotResult</a></span><span class="op">(</span><span class="va">spe_query</span><span class="op">$</span><span class="va">pred_set</span><span class="op">[[</span><span class="fl">75</span><span class="op">]</span><span class="op">]</span>, <span class="va">opi2</span>,</span>
<span>    probs <span class="op">=</span> <span class="va">p_test</span><span class="op">[</span><span class="fl">75</span>, <span class="op">]</span>,</span>
<span>    col_grad <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"lemonchiffon"</span>, <span class="st">"orange"</span>, <span class="st">"darkred"</span><span class="op">)</span>,</span>
<span>    attrs <span class="op">=</span> <span class="va">attrs</span>, add_scores <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    title <span class="op">=</span> <span class="st">"Conformal Prediction set"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="scConform_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotResult.html">plotResult</a></span><span class="op">(</span><span class="va">spe_query</span><span class="op">$</span><span class="va">pred_set_hier</span><span class="op">[[</span><span class="fl">75</span><span class="op">]</span><span class="op">]</span>, <span class="va">opi2</span>,</span>
<span>    probs <span class="op">=</span> <span class="va">p_test</span><span class="op">[</span><span class="fl">75</span>, <span class="op">]</span>,</span>
<span>    col_grad <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"lemonchiffon"</span>, <span class="st">"orange"</span>, <span class="st">"darkred"</span><span class="op">)</span>,</span>
<span>    attrs <span class="op">=</span> <span class="va">attrs</span>, add_scores <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    title <span class="op">=</span> <span class="st">"Hierarchical Prediction set"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="scConform_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="r-session-info">R.session Info<a class="anchor" aria-label="anchor" href="#r-session-info"></a>
</h2>
<pre><code>R version 4.4.1 (2024-06-14)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] splines   stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] BiocParallel_1.38.0         scran_1.32.0               
 [3] scuttle_1.14.0              igraph_2.0.3               
 [5] MerfishData_1.6.0           SpatialExperiment_1.14.0   
 [7] EBImage_4.46.0              ontoProc_1.26.0            
 [9] ontologyIndex_2.12          VGAM_1.1-11                
[11] SingleCellExperiment_1.26.0 SummarizedExperiment_1.34.0
[13] Biobase_2.64.0              GenomicRanges_1.56.1       
[15] GenomeInfoDb_1.40.1         IRanges_2.38.0             
[17] S4Vectors_0.42.0            BiocGenerics_0.50.0        
[19] MatrixGenerics_1.16.0       matrixStats_1.3.0          
[21] scConform_0.99.1            BiocStyle_2.32.1           

loaded via a namespace (and not attached):
  [1] jsonlite_1.8.8            magrittr_2.0.3           
  [3] magick_2.8.3              rmarkdown_2.27           
  [5] fs_1.6.4                  zlibbioc_1.50.0          
  [7] ragg_1.3.2                vctrs_0.6.5              
  [9] memoise_2.0.1             DelayedMatrixStats_1.26.0
 [11] RCurl_1.98-1.14           htmltools_0.5.8.1        
 [13] S4Arrays_1.4.1            AnnotationHub_3.12.0     
 [15] curl_5.2.1                BiocNeighbors_1.22.0     
 [17] SparseArray_1.4.8         sass_0.4.9               
 [19] bslib_0.7.0               htmlwidgets_1.6.4        
 [21] desc_1.4.3                cachem_1.1.0             
 [23] mime_0.12                 lifecycle_1.0.4          
 [25] pkgconfig_2.0.3           rsvd_1.0.5               
 [27] Matrix_1.7-0              R6_2.5.1                 
 [29] fastmap_1.2.0             GenomeInfoDbData_1.2.12  
 [31] shiny_1.8.1.1             digest_0.6.36            
 [33] AnnotationDbi_1.66.0      dqrng_0.4.1              
 [35] irlba_2.3.5.1             ExperimentHub_2.12.0     
 [37] textshaping_0.4.0         RSQLite_2.3.7            
 [39] beachmat_2.20.0           filelock_1.0.3           
 [41] fansi_1.0.6               httr_1.4.7               
 [43] abind_1.4-5               compiler_4.4.1           
 [45] withr_3.0.0               bit64_4.0.5              
 [47] tiff_0.1-12               DBI_1.2.3                
 [49] highr_0.11                rappdirs_0.3.3           
 [51] DelayedArray_0.30.1       rjson_0.2.21             
 [53] bluster_1.14.0            tools_4.4.1              
 [55] httpuv_1.6.15             glue_1.7.0               
 [57] promises_1.3.0            grid_4.4.1               
 [59] cluster_2.1.6             generics_0.1.3           
 [61] metapod_1.12.0            BiocSingular_1.20.0      
 [63] ScaledMatrix_1.12.0       utf8_1.2.4               
 [65] XVector_0.44.0            BiocVersion_3.19.1       
 [67] pillar_1.9.0              limma_3.60.3             
 [69] BumpyMatrix_1.12.0        later_1.3.2              
 [71] dplyr_1.1.4               BiocFileCache_2.12.0     
 [73] lattice_0.22-6            bit_4.0.5                
 [75] tidyselect_1.2.1          locfit_1.5-9.10          
 [77] Biostrings_2.72.1         knitr_1.47               
 [79] bookdown_0.39             edgeR_4.2.0              
 [81] ontologyPlot_1.7          xfun_0.45                
 [83] statmod_1.5.0             DT_0.33                  
 [85] UCSC.utils_1.0.0          fftwtools_0.9-11         
 [87] paintmap_1.0              yaml_2.3.8               
 [89] evaluate_0.24.0           codetools_0.2-20         
 [91] tibble_3.2.1              Rgraphviz_2.48.0         
 [93] BiocManager_1.30.23       graph_1.82.0             
 [95] cli_3.6.3                 xtable_1.8-4             
 [97] reticulate_1.38.0         systemfonts_1.1.0        
 [99] jquerylib_0.1.4           Rcpp_1.0.12              
[101] dbplyr_2.5.0              png_0.1-8                
[103] parallel_4.4.1            pkgdown_2.0.9            
[105] blob_1.2.4                jpeg_0.1-10              
[107] sparseMatrixStats_1.16.0  bitops_1.0-7             
[109] purrr_1.0.2               crayon_1.5.3             
[111] rlang_1.1.4               KEGGREST_1.44.1          </code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Corbetta Daniela.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
